integrations/benchmarks_test.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package integrations
+
+import (
+	"fmt"
+	"math/rand"
+	"net/http"
+	"testing"
+
+	"code.gitea.io/gitea/models"
+	api "code.gitea.io/sdk/gitea"
+)
+
+func BenchmarkRepo(b *testing.B) {
+	samples := []struct {
+		url       string
+		name      string
+		skipShort bool
+	}{
+		{url: "https://github.com/go-gitea/gitea.git", name: "gitea"},
+		{url: "https://github.com/ethantkoenig/manyfiles.git", name: "manyfiles"},
+		{url: "https://github.com/moby/moby.git", name: "moby", skipShort: true},
+		{url: "https://github.com/golang/go.git", name: "go", skipShort: true},
+		{url: "https://github.com/torvalds/linux.git", name: "linux", skipShort: true},
+	}
+	prepareTestEnv(b)
+	session := loginUser(b, "user2")
+	b.ResetTimer()
+
+	for _, s := range samples {
+		b.Run(s.name, func(b *testing.B) {
+			if testing.Short() && s.skipShort {
+				b.Skip("skipping test in short mode.")
+			}
+			b.Run("Migrate", func(b *testing.B) {
+				for i := 0; i < b.N; i++ {
+					testRepoMigrate(b, session, s.url, s.name)
+				}
+			})
+			b.Run("Access", func(b *testing.B) {
+				var branches []*api.Branch
+				b.Run("APIBranchList", func(b *testing.B) {
+					for i := 0; i < b.N; i++ {
+						req := NewRequestf(b, "GET", "/api/v1/repos/%s/%s/branches", "user2", s.name)
+						resp := session.MakeRequest(b, req, http.StatusOK)
+						b.StopTimer()
+						if len(branches) == 0 {
+							DecodeJSON(b, resp, &branches) //Store for next phase
+						}
+						b.StartTimer()
+					}
+				})
+				branchCount := len(branches)
+				b.Run("WebBranchCommit", func(b *testing.B) {
+					for i := 0; i < b.N; i++ {
+						req := NewRequestf(b, "GET", "/%s/%s/commit/%s", "user2", s.name, branches[i%branchCount].Commit.ID)
+						session.MakeRequest(b, req, http.StatusOK)
+					}
+				})
+			})
+		})
+	}
+}
+
+//StringWithCharset random string (from https://www.calhoun.io/creating-random-strings-in-go/)
+func StringWithCharset(length int, charset string) string {
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[rand.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+func BenchmarkRepoBranchCommit(b *testing.B) {
+	samples := []struct {
+		repoID int64
+		repo   *models.Repository
+		owner  *models.User
+	}{
+		{repoID: 1},
+		{repoID: 3},
+		{repoID: 15},
+		{repoID: 16},
+	}
+
+	prepareTestEnv(b)
+	for i, s := range samples {
+		samples[i].repo = models.AssertExistsAndLoadBean(b, &models.Repository{ID: s.repoID}).(*models.Repository)
+		samples[i].owner = models.AssertExistsAndLoadBean(b, &models.User{ID: samples[i].repo.OwnerID}).(*models.User)
+	}
```

Comment:
I think this can be simplified. Just make `samples` a list of repo IDs, and load the owner and repo in the main loop. We can get rid of the anonymous struct and this first loop.

```
for _, repoID := range repoIDs {
	repo := models.AssertExistsAndLoadBean(b, &models.Repository{ID: repoID}).(*models.Repository)
	owner := models.AssertExistsAndLoadBean(b, &models.User{ID: repo.OwnerID}).(*models.User)
	... // run test
}
```